### DISCUSSION TITLE
need some support for understanding the code

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22050] [Q&A General] [answered]
Hello smart guys, 

I just look at some codes and want to get some help from you to understand it better (see attached figure).
Could you breifly describe what is going on with p0, p1 and different cases? Thank you.

Kind regards,
Qia

**Comment by GiudGiud** [comment, marked as answer]
Howdy

Please stop posting screenshots, this is text, it can be posted in a ```code snippet```

direction can be X, -X, Y etc
depending on this choice, the definition of the vector from p0 to p1 is chosen different based on how p0 and p1 are ordered along the axis

Guillaume

**Reply by Oops-Qiao** [reply]
Howdy，

Ok，I will use code snippet in the future. Thank you. :)
But just to confirm again,   what does it mean with p0 = _current_elem->point(0);  p1 = _current_elem->point(1).

**Reply by GiudGiud** [reply]
(0) or (1) retrieves the coordinate along the X and Y axis respectively

**Reply by Oops-Qiao** [reply]
umm, well, if (0) or (1) retrieves the coordinate along the X and Y axis respectively, in case 1, he wrote p=p0-p1 which means a X coordinate-Y coordinate? how to understand this?
By the way, from case 1 to case 6, how to understand p0(0), p0(1) and p0(2)?
I really appreciate your help Guillaume. Sorry, I am a beginner.

**Reply by GiudGiud** [reply]
well they're initializing p0, a vector with a point(0) so it assigns the value to all coordinates of that vector.

then the operations on p0 and p1 (p0-p1 for example) are vector operations as well.

**Reply by GiudGiud** [reply]
I dont know how to interpret that particular code. You can try different inputs and look at the outputs. What is this for?

**Reply by Oops-Qiao** [reply]
Thank you very much and I really appreciate the patience from you and your team. Thank you Guillaume.

**Reply by GiudGiud** [reply]
The point class is defined in libmesh not moose. So you ll have to take a look at the libmesh doxygen

**Reply by Oops-Qiao** [reply]
Ok, I see. Thank you very much Guillaume.

**Comment by dschwen** [comment]
who wrote this code? the assignment inside the ternary is rather unusual. I suggest pulling out the `p = ` to get 

```
const int i = (direction / 2) + 1;
const int d = direction % 2;
p = (p0(i) < p1(i)) == (d == 0) ? p0 - p1 : p1 - p0;
```

**Reply by hugary1995** [reply]
I agree with @dschwen . @Oops-Qiao you've mentioned several times that you are just beginning to learn MOOSE. I think learning MOOSE from code in an external app that hasn't been peer-reviewed isn't your best option. Based on your questions, I feel like your best starting point is probably to go through the C++ fundamentals https://mooseframework.inl.gov/help/c++/index.html. You can also dig into the doxygen documentations on this page https://mooseframework.inl.gov/framework_development/index.html which will answer most of the questions you answered.

**Reply by Oops-Qiao** [reply]
Yeah, thanks a lot for the links Tianchen, that is really helpful to me.
